src/dbnode/integration/bootstrap_retries_test.go

```diff
@@ -103,35 +106,115 @@ func TestBootstrapRetriesDueToObsoleteRanges(t *testing.T) {
-		require.NoError(t, setup.WaitUntilServerIsUp())
+	go assertBootstrapRetry(t, setup, signalCh)
 
-		// First bootstrap pass, persist ranges. Check if DB is not marked bootstrapped and advance clock.
-		signalCh <- struct{}{}
-		assert.False(t, setup.DB().IsBootstrapped(), "database should not yet be bootstrapped")
-		setup.SetNowFn(setup.NowFn()().Add(2 * time.Hour))
-		signalCh <- struct{}{}
+	require.NoError(t, setup.StartServer()) // Blocks until bootstrap is complete
+	defer func() {
+		require.NoError(t, setup.StopServer())
+	}()
 
-		// Still first bootstrap pass, in-memory ranges. Due to advanced clock previously calculated
-		// ranges are obsolete. Check if DB is not marked bootstrapped.
-		signalCh <- struct{}{}
-		assert.False(t, setup.DB().IsBootstrapped(), "database should not yet be bootstrapped")
-		signalCh <- struct{}{}
+	assert.True(t, setup.DB().IsBootstrapped(), "database should be bootstrapped")
+	assertRetryMetric(t, testScope, "obsolete-ranges")
+}
 
-		// Bootstrap retry, persist ranges. Check if DB isn't marked as bootstrapped on the second pass.
-		signalCh <- struct{}{}
-		assert.False(t, setup.DB().IsBootstrapped(), "database should not yet be bootstrapped")
-		signalCh <- struct{}{}
+func TestNoOpenFilesWhenBootstrapRetriesDueToObsoleteRanges(t *testing.T) {
+	// Setup the test bootstrapper to only proceed when a signal is sent.
+	signalCh := make(chan struct{})
 
-		// Still bootstrap retry, in-memory ranges. DB finishes bootstrapping.
-		signalCh <- struct{}{}
-		assert.False(t, setup.DB().IsBootstrapped(), "database should not yet be bootstrapped")
-		signalCh <- struct{}{}
-	}()
+	setup, testScope := bootstrapRetryTestSetup(t, func(
+		ctx context.Context,
+		namespaces bootstrap.Namespaces,
+		cache bootstrap.Cache,
+	) (bootstrap.NamespaceResults, error) {
+		// read from signalCh twice so we could advance the clock exactly in between of those signals
+		<-signalCh
+		<-signalCh
+		bs, err := bootstrapper.NewNoOpAllBootstrapperProvider().Provide()
+		require.NoError(t, err)
+		return bs.Bootstrap(ctx, namespaces, cache)
+	})
+
+	go assertBootstrapRetry(t, setup, signalCh)
+
+	// Write test data
+	now := setup.NowFn()()
+
+	fooSeries := generate.Series{
+		ID:   ident.StringID("foo"),
+		Tags: ident.NewTags(ident.StringTag("city", "new_york"), ident.StringTag("foo", "foo")),
+	}
+
+	barSeries := generate.Series{
+		ID:   ident.StringID("bar"),
+		Tags: ident.NewTags(ident.StringTag("city", "new_jersey")),
+	}
 
+	bazSeries := generate.Series{
+		ID:   ident.StringID("baz"),
+		Tags: ident.NewTags(ident.StringTag("city", "seattle")),
+	}
+	var (
+		blockSize = 2 * time.Hour
+		rOpts     = retention.NewOptions().SetRetentionPeriod(6 * blockSize).SetBlockSize(blockSize)
+		idxOpts   = namespace.NewIndexOptions().SetEnabled(true).SetBlockSize(2 * blockSize)
+		nOpts     = namespace.NewOptions().SetRetentionOptions(rOpts).SetIndexOptions(idxOpts)
+	)
+	ns1, err := namespace.NewMetadata(testNamespaces[0], nOpts)
```

Comment:
You should be able to retrieve namespaces from `setup.Namespaces()`

https://github.com/m3db/m3/blob/255a4b14df1da9b4bc702814933bd8b15d0b3caf/src/dbnode/integration/setup.go#L156